File: examples/dsp-library/docs/snippets/cross-language-contract.mjs
import assert from 'node:assert/strict';
import fs from 'node:fs';

import * as dsp from '@effetune/dsp';

const python = JSON.parse(fs.readFileSync(process.argv[2], 'utf8'));
const types = dsp.EFFECT_CATALOG.effects.map(effect => effect.type);
assert.equal(types.length, 102);
assert.deepEqual(dsp.EFFECT_CATALOG, python.catalog);
assert.deepEqual(dsp.EFFECT_CATALOG.channels, dsp.EFFECT_CHANNELS);
assert.ok(dsp.EFFECT_CATALOG.effects.every(effect =>
  !Object.hasOwn(effect, 'implementation')
));
assert.deepEqual([...dsp.EFFECT_CLASSES.keys()], types);
for (const type of types) {
  assert.equal(typeof dsp[type], 'function', `${type} class is not a root export.`);
  assert.equal(
    typeof dsp[`create${type}`],
    'function',
    `create${type} factory is not a root export.`
  );
}

const chain = await dsp.createChain([
  new dsp.BrickwallLimiter({ id: 'limiter', lookahead: 3 }),
  new dsp.BrickwallLimiter({ id: 'disabled', enabled: false, lookahead: 10 })
], { variant: 'baseline' });
const stream = await chain.stream({
  sampleRate: 48000,
  channels: 2,
  blockSize: 64,
  seed: 0
});
assert.equal(stream.latencySamples, python.latencySamples);
stream.close();
chain.close();

File: examples/dsp-library/docs/snippets/cross-language-contract.py
import json
import sys

import effetune as et


types = [effect["type"] for effect in et.EFFECT_METADATA["effects"]]
assert len(types) == 102
assert list(et.EFFECT_CLASSES) == types
assert all(getattr(et, effect_type) is et.EFFECT_CLASSES[effect_type] for effect_type in types)
assert all("implementation" not in effect for effect in et.EFFECT_METADATA["effects"])

chain = et.Chain([
    et.BrickwallLimiter(id="limiter", lookahead=3),
    et.BrickwallLimiter(id="disabled", enabled=False, lookahead=10),
])
with chain.stream(48_000, channels=2, block_size=64) as stream:
    latency_samples = stream.latency_samples

destination = sys.argv[1] if len(sys.argv) > 1 else "python-contract.json"
with open(destination, "w", encoding="utf-8", newline="\n") as output:
    json.dump(
        {
            "catalog": et.EFFECT_METADATA,
            "latencySamples": latency_samples,
        },
        output,
        ensure_ascii=False,
        sort_keys=True,
    )
    output.write("\n")
